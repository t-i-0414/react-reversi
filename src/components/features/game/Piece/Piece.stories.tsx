import { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import { Piece } from './Piece';

const meta: Meta<typeof Piece> = {
  title: 'features/game/Piece',
  component: Piece,
  argTypes: {
    pieceColor: {
      control: {
        type: 'select',
        options: {
          white: 'white',
          black: 'black',
          canTurnOver: 'canTurnOver',
        },
      },
    },
    size: { control: { type: 'text' } },
    onclick: { action: 'clicked' },
    dataCy: { control: { type: 'text' } },
  },
  args: {
    pieceColor: 'white',
    size: '64px',
    dataCy: 'dataCY',
  },
};

export default meta;

type Story = StoryObj<typeof Piece>;

export const White: Story = {
  args: { pieceColor: 'white' },
  render: props => <Piece {...props} />,
};

export const Black: Story = {
  args: { pieceColor: 'black' },
  render: props => <Piece {...props} />,
};

export const CanTurnOver: Story = {
  args: { pieceColor: 'canTurnOver' },
  render: props => <Piece {...props} />,
};
