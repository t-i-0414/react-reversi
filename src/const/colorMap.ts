export const ColorMap = {
  // common colors
  WHITE: 'fff',

  // text colors
  TX_DEEP_BLACK: '#000',
  TX_BLACK: '#555',
  TX_WHITE: '#fff',
  TX_RED: '#F62043',

  // background colors
  BG_GREEN: '#38A34B',
  BG_WHITE: '#fff',
  BG_LIGHT_BLACK: '#262626',
  BG_BLACK: '#000',
  BG_BLACK_ALPHA30: 'rgba(0, 0, 0, 0.3)',
  BG_LIGHT_GRAY: '#c7c7c7',

  // border colors
  BD_BLACK: '#000',
  BD_LIGHT_GRAY: '#c7c7c7',

  // piece colors
  PC_WHITE: '#fff',
  PC_BLACK: '#000',
  PC_INVISIBLE: '#38A34B',
} as const;
