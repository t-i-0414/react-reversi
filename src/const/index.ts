export { ColorMap } from '~/const/colorMap';
export { PieceColor } from '~/const/pieceColorMap';
export { PlayerMap } from '~/const/playerMap';
export { SizeMap } from '~/const/sizeMap';
