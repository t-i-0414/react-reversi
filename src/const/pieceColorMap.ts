import type { AllPieceColorMap } from '~/types';

export const PieceColor: AllPieceColorMap = {
  WHITE: 'white',
  BLACK: 'black',
  INVISIBLE: 'invisible',
} as const;
