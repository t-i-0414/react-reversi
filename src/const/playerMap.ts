export const PlayerMap = {
  PLAYER_1: {
    name: 'Player 1',
    shortName: 'P1',
    playerValue: '1',
  },
  PLAYER_2: {
    name: 'Player 2',
    shortName: 'P2',
    playerValue: '2',
  },
  NONE: {
    name: 'NONE',
    shortName: 'NONE',
    playerValue: '0',
  },
} as const;
