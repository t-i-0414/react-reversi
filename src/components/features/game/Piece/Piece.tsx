import React from 'react';
import styled, { css } from 'styled-components';
import { ColorMap } from '~/const';
import type { Piece as TPiece } from '~/domains';

export interface Props {
  pieceColor: TPiece;
  size?: string;
  onclick?: () => void;
  dataCy?: string;
}

export const Piece: React.FC<Props> = ({
  pieceColor,
  size = '64px',
  onclick,
  dataCy,
}) => {
  switch (pieceColor) {
    case 'white':
      return <StyledPiece color={ColorMap.PC_WHITE} size={size} />;
    case 'black':
      return <StyledPiece color={ColorMap.PC_BLACK} size={size} />;
    case 'canTurnOver':
      return (
        <StyledPiece
          color={ColorMap.PC_INVISIBLE}
          size={size}
          onClick={onclick}
          data-cy={dataCy}
        />
      );
    default: {
      const exhaustiveCheck: never = pieceColor;

      return exhaustiveCheck;
    }
  }
};

type StyledPieceProps = {
  color: (typeof ColorMap)[keyof Pick<
    typeof ColorMap,
    'PC_WHITE' | 'PC_BLACK' | 'PC_INVISIBLE'
  >];
  size: string;
};
const StyledPiece = styled.span<StyledPieceProps>`
  box-sizing: border-box;
  display: block;
  width: ${props => props.size};
  height: ${props => props.size};
  background-color: ${props => props.color};
  ${props =>
    props.color === ColorMap.PC_INVISIBLE
      ? css`
          border: none;
          border-radius: 50%;

          &:hover {
            cursor: pointer;
          }
        `
      : css`
          border: 1px solid ${ColorMap.BD_BLACK};
          border-radius: 50%;
        `}
`;
