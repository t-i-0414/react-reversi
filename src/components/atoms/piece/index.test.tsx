import { expect } from '@jest/globals';
import React from 'react';
import renderer from 'react-test-renderer';
import { PieceColor } from '~/const';
import Piece from '.';

describe('piece unit test', () => {
  it('snapshot(white)', () => {
    const component = renderer.create(<Piece pieceColor={PieceColor.WHITE} />);
    expect(component).toMatchSnapshot();
  });
  it('snapshot(black)', () => {
    const component = renderer.create(<Piece pieceColor={PieceColor.BLACK} />);
    expect(component).toMatchSnapshot();
  });
  it('snapshot(clickable)', () => {
    const component = renderer.create(
      <Piece
        pieceColor={PieceColor.INVISIBLE}
        onclick={() => 'This is clickable Piece.'}
        dataCy='clickable'
      />,
    );
    expect(component).toMatchSnapshot();
  });
});
