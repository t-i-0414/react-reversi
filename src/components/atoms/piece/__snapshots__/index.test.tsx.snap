// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`piece unit test snapshot(black) 1`] = `
<span
  className="sc-bcXHqe JfklG"
  color="#000"
/>
`;

exports[`piece unit test snapshot(clickable) 1`] = `
<span
  className="sc-bcXHqe fSVNNw"
  color="#38A34B"
  data-cy="clickable"
  onClick={[Function]}
/>
`;

exports[`piece unit test snapshot(white) 1`] = `
<span
  className="sc-bcXHqe dJWbKk"
  color="#fff"
/>
`;
